PR failed `sandbox` IT due to #7660, but also the failed (mirroring) queues weren't empty.

```
======================================================================
FAIL: test_indexing (integration_test.IndexingIntegrationTest.test_indexing) [mirror_sources_and_files]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builds/ucsc/azul/test/integration_test.py", line 385, in subTest
    yield
  File "/builds/ucsc/azul/test/integration_test.py", line 1789, in _test_mirroring
    self._assert_queues_empty([config.mirror_queue.to_fail.name])
    ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/builds/ucsc/azul/test/integration_test.py", line 1410, in _assert_queues_empty
    self.assertTrue(self.azul_client.is_queue_empty(queue_name))
    ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: False is not true
----------------------------------------------------------------------
```

Failing section seems to have been introduced in this PR.

**CloudWatch Logs Insights**    
region: us-east-1    
log-group-names: /aws/lambda/azul-indexer-sandbox-mirror    
data-sources:     
facets:     
start-time: 2026-01-06T07:30:05.112Z    
end-time: 2026-01-06T07:31:51.882Z    
query-string:
  ```
  fields @timestamp, @message
| filter @message like /ERROR/
| sort @timestamp desc
| limit 10000
  ```
---
| @timestamp | @message |
| --- | --- |
| 2026-01-06 07:31:05.051 | [ERROR] AssertionError Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;chalice.py&quot;, line 194, in patched_event_source_handler     return old_handler(self_, event, context)   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1761, in __call__     return self.handler(event_obj)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 65, in mirror     self.mirror(event)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 75, in mirror     self._handle_events(event, message_handler)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;action_controller.py&quot;, line 50, in _handle_events     message_handler(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 73, in message_handler     self.service(action.catalog).mirror(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 539, in mirror     self._queue_actions(self._mirror(action))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 408, in _queue_actions     return self._queues.send_messages(self._mirror_queue(),   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;queues.py&quot;, line 208, in send_messages     for batch in chunked(messages, self.batch_size):   File &quot;&#x2f;opt&#x2f;python&#x2f;more_itertools&#x2f;recipes.py&quot;, line 133, in take     return list(islice(iterable, n))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 663, in _     yield devolve(FinalizeFileAction, a, upload=upload)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;attrs.py&quot;, line 951, in devolve     assert issubclass(subcls, cls) |
| 2026-01-06 07:31:00.479 | [ERROR] AssertionError Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;chalice.py&quot;, line 194, in patched_event_source_handler     return old_handler(self_, event, context)   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1761, in __call__     return self.handler(event_obj)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 65, in mirror     self.mirror(event)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 75, in mirror     self._handle_events(event, message_handler)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;action_controller.py&quot;, line 50, in _handle_events     message_handler(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 73, in message_handler     self.service(action.catalog).mirror(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 539, in mirror     self._queue_actions(self._mirror(action))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 408, in _queue_actions     return self._queues.send_messages(self._mirror_queue(),   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;queues.py&quot;, line 208, in send_messages     for batch in chunked(messages, self.batch_size):   File &quot;&#x2f;opt&#x2f;python&#x2f;more_itertools&#x2f;recipes.py&quot;, line 133, in take     return list(islice(iterable, n))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 663, in _     yield devolve(FinalizeFileAction, a, upload=upload)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;attrs.py&quot;, line 951, in devolve     assert issubclass(subcls, cls) |
| 2026-01-06 07:30:57.004 | [ERROR] AssertionError Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;chalice.py&quot;, line 194, in patched_event_source_handler     return old_handler(self_, event, context)   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1761, in __call__     return self.handler(event_obj)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 65, in mirror     self.mirror(event)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 75, in mirror     self._handle_events(event, message_handler)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;action_controller.py&quot;, line 50, in _handle_events     message_handler(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 73, in message_handler     self.service(action.catalog).mirror(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 539, in mirror     self._queue_actions(self._mirror(action))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 408, in _queue_actions     return self._queues.send_messages(self._mirror_queue(),   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;queues.py&quot;, line 208, in send_messages     for batch in chunked(messages, self.batch_size):   File &quot;&#x2f;opt&#x2f;python&#x2f;more_itertools&#x2f;recipes.py&quot;, line 133, in take     return list(islice(iterable, n))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 663, in _     yield devolve(FinalizeFileAction, a, upload=upload)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;attrs.py&quot;, line 951, in devolve     assert issubclass(subcls, cls) |
| 2026-01-06 07:30:56.909 | [ERROR] AssertionError Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;chalice.py&quot;, line 194, in patched_event_source_handler     return old_handler(self_, event, context)   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1761, in __call__     return self.handler(event_obj)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 65, in mirror     self.mirror(event)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 75, in mirror     self._handle_events(event, message_handler)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;action_controller.py&quot;, line 50, in _handle_events     message_handler(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 73, in message_handler     self.service(action.catalog).mirror(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 539, in mirror     self._queue_actions(self._mirror(action))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 408, in _queue_actions     return self._queues.send_messages(self._mirror_queue(),   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;queues.py&quot;, line 208, in send_messages     for batch in chunked(messages, self.batch_size):   File &quot;&#x2f;opt&#x2f;python&#x2f;more_itertools&#x2f;recipes.py&quot;, line 133, in take     return list(islice(iterable, n))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 663, in _     yield devolve(FinalizeFileAction, a, upload=upload)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;attrs.py&quot;, line 951, in devolve     assert issubclass(subcls, cls) |
| 2026-01-06 07:30:51.435 | [ERROR] AssertionError Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;chalice.py&quot;, line 194, in patched_event_source_handler     return old_handler(self_, event, context)   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1761, in __call__     return self.handler(event_obj)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 65, in mirror     self.mirror(event)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 75, in mirror     self._handle_events(event, message_handler)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;action_controller.py&quot;, line 50, in _handle_events     message_handler(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 73, in message_handler     self.service(action.catalog).mirror(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 539, in mirror     self._queue_actions(self._mirror(action))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 408, in _queue_actions     return self._queues.send_messages(self._mirror_queue(),   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;queues.py&quot;, line 208, in send_messages     for batch in chunked(messages, self.batch_size):   File &quot;&#x2f;opt&#x2f;python&#x2f;more_itertools&#x2f;recipes.py&quot;, line 133, in take     return list(islice(iterable, n))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 663, in _     yield devolve(FinalizeFileAction, a, upload=upload)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;attrs.py&quot;, line 951, in devolve     assert issubclass(subcls, cls) |
| 2026-01-06 07:30:51.336 | [ERROR] AssertionError Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;chalice.py&quot;, line 194, in patched_event_source_handler     return old_handler(self_, event, context)   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1761, in __call__     return self.handler(event_obj)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 65, in mirror     self.mirror(event)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 75, in mirror     self._handle_events(event, message_handler)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;action_controller.py&quot;, line 50, in _handle_events     message_handler(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 73, in message_handler     self.service(action.catalog).mirror(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 539, in mirror     self._queue_actions(self._mirror(action))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 408, in _queue_actions     return self._queues.send_messages(self._mirror_queue(),   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;queues.py&quot;, line 208, in send_messages     for batch in chunked(messages, self.batch_size):   File &quot;&#x2f;opt&#x2f;python&#x2f;more_itertools&#x2f;recipes.py&quot;, line 133, in take     return list(islice(iterable, n))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 663, in _     yield devolve(FinalizeFileAction, a, upload=upload)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;attrs.py&quot;, line 951, in devolve     assert issubclass(subcls, cls) |
| 2026-01-06 07:30:47.889 | [ERROR] AssertionError Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;chalice.py&quot;, line 194, in patched_event_source_handler     return old_handler(self_, event, context)   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1761, in __call__     return self.handler(event_obj)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 65, in mirror     self.mirror(event)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 75, in mirror     self._handle_events(event, message_handler)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;action_controller.py&quot;, line 50, in _handle_events     message_handler(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 73, in message_handler     self.service(action.catalog).mirror(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 539, in mirror     self._queue_actions(self._mirror(action))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 408, in _queue_actions     return self._queues.send_messages(self._mirror_queue(),   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;queues.py&quot;, line 208, in send_messages     for batch in chunked(messages, self.batch_size):   File &quot;&#x2f;opt&#x2f;python&#x2f;more_itertools&#x2f;recipes.py&quot;, line 133, in take     return list(islice(iterable, n))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 663, in _     yield devolve(FinalizeFileAction, a, upload=upload)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;attrs.py&quot;, line 951, in devolve     assert issubclass(subcls, cls) |
| 2026-01-06 07:30:40.944 | [ERROR] AssertionError Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;chalice.py&quot;, line 194, in patched_event_source_handler     return old_handler(self_, event, context)   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1761, in __call__     return self.handler(event_obj)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 65, in mirror     self.mirror(event)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 75, in mirror     self._handle_events(event, message_handler)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;action_controller.py&quot;, line 50, in _handle_events     message_handler(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 73, in message_handler     self.service(action.catalog).mirror(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 539, in mirror     self._queue_actions(self._mirror(action))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 408, in _queue_actions     return self._queues.send_messages(self._mirror_queue(),   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;queues.py&quot;, line 208, in send_messages     for batch in chunked(messages, self.batch_size):   File &quot;&#x2f;opt&#x2f;python&#x2f;more_itertools&#x2f;recipes.py&quot;, line 133, in take     return list(islice(iterable, n))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 663, in _     yield devolve(FinalizeFileAction, a, upload=upload)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;attrs.py&quot;, line 951, in devolve     assert issubclass(subcls, cls) |
---

--- a/.github/PULL_REQUEST_TEMPLATE/anvilprod-hotfix.md
+++ b/.github/PULL_REQUEST_TEMPLATE/anvilprod-hotfix.md
@@ -59,11 +59,6 @@ Linked issue: #0000
 - [ ] Pushed PR branch to GitHub
 
 
-### Operator (deploy runner image)
-
-- [ ] Ran `_select anvilprod.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
-
-
 ### Operator (sandbox build)
 
 - [ ] Added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>

--- a/.github/PULL_REQUEST_TEMPLATE/backport.md
+++ b/.github/PULL_REQUEST_TEMPLATE/backport.md
@@ -46,12 +46,6 @@ This is the PR template for backport PRs against `develop`.
 - [ ] Pushed PR branch to GitHub
 
 
-### Operator (deploy runner image)
-
-- [ ] Ran `_select dev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
-- [ ] Ran `_select anvildev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
-
-
 ### Operator (sandbox build)
 
 - [ ] Added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>

--- a/.github/PULL_REQUEST_TEMPLATE/prod-hotfix.md
+++ b/.github/PULL_REQUEST_TEMPLATE/prod-hotfix.md
@@ -59,11 +59,6 @@ Linked issue: #0000
 - [ ] Pushed PR branch to GitHub
 
 
-### Operator (deploy runner image)
-
-- [ ] Ran `_select prod.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
-
-
 ### Operator (merge the branch)
 
 - [ ] All status checks passed and the PR is mergeable

--- a/.github/pull_request_template.md.template.py
+++ b/.github/pull_request_template.md.template.py
@@ -798,21 +798,23 @@ def emit(t: T, target_branch: str):
                     'content': 'PR is assigned to only the operator and the author',
                 },
             ]),
-            {
-                'type': 'h2',
-                'content': 'Operator (deploy runner image)'
-            },
-            *[
+            *iif(t not in (T.hotfix, T.backport), [
                 {
-                    'type': 'cli',
-                    'content': 'Ran ' + bq(
-                        f'_select {d}.gitlab && '
-                        f'make -C terraform/gitlab/runner'
-                    ),
-                    'alt': 'or this PR is not labeled `deploy:runner`'
-                }
-                for d in t.target_deployments(target_branch)
-            ],
+                    'type': 'h2',
+                    'content': 'Operator (deploy runner image)'
+                },
+                *[
+                    {
+                        'type': 'cli',
+                        'content': 'Ran ' + bq(
+                            f'_select {d}.gitlab && '
+                            f'make -C terraform/gitlab/runner'
+                        ),
+                        'alt': 'or this PR is not labeled `deploy:runner`'
+                    }
+                    for d in t.target_deployments(target_branch)
+                ]
+            ]),
             *iif(t.has_sandbox_for(target_branch), [
                 {
                     'type': 'h2',

--- a/src/azul/indexer/action_controller.py
+++ b/src/azul/indexer/action_controller.py
@@ -41,16 +41,9 @@ class ActionController[A: Action](AppController, metaclass=ABCMeta):
                        event: Iterable[SQSRecord],
                        message_handler: Callable[[A], None]):
         for record in event:
-            message: SQSMessage
-            if self.actions_are_fifo:
-                message = SQSFifoMessage.from_record(record)
-                suffix, args = ', group ID %s', [message.group_id]
-            else:
-                message = SQSMessage.from_record(record)
-                suffix, args = '', []
-            log.info('Worker handling message %r, ' +
-                     'attempt #%i (approx), message ID %s' + suffix,
-                     message.body, message.attempts, message.id, *args)
+            message_cls = SQSFifoMessage if self.actions_are_fifo else SQSMessage
+            message = message_cls.from_record(record)
+            log.info('Worker handling message %r', message)
             start = time.time()
             try:
                 action = self.action_cls.from_json(message.body)

--- a/src/azul/indexer/index_queue_service.py
+++ b/src/azul/indexer/index_queue_service.py
@@ -118,7 +118,6 @@ class IndexQueueService:
                            ) -> None:
         queue = self.notifications_queue(retry=retry)
         self.queues.send_message(queue, message)
-        log.info('Queued notification message %r', message)
 
     def queue_tallies(self,
                       messages: Iterable[SQSMessage],
@@ -126,7 +125,8 @@ class IndexQueueService:
                       retry: bool = False
                       ) -> int:
         queue = self.tallies_queue(retry=retry)
-        return self.queues.send_messages(queue, messages)
+        # Logging tallies would be excessively verbose
+        return self.queues.send_messages(queue, messages, log_level=0)
 
     def index_bundle_message(self,
                              catalog: CatalogName,

--- a/src/azul/indexer/mirror_service.py
+++ b/src/azul/indexer/mirror_service.py
@@ -304,7 +304,7 @@ class BaseMirrorService:
 
     def mirror_sources(self, sources: Iterable[tuple[SourceRef, SourceConfig]]):
         if self.may_mirror():
-            def messages():
+            def actions():
                 for source, source_config in sources:
                     if source_config.mirror:
                         log.info('Mirroring files in source %r from catalog %r',
@@ -315,27 +315,27 @@ class BaseMirrorService:
                                  'mirroring is explicitly disabled',
                                  str(source.spec), self.catalog)
 
-            self._queue_messages(messages())
+            self._queue_actions(actions())
         else:
             log.info('Not mirroring any files in catalog %r because the file '
                      'size limit is negative', self.catalog)
 
     def mirror_file(self, source: SourceRef, file: File):
-        self._queue_messages([MirrorFileAction(catalog=self.catalog,
-                                               source=source,
-                                               prefix='',
-                                               file=file)])
+        self._queue_actions([MirrorFileAction(catalog=self.catalog,
+                                              source=source,
+                                              prefix='',
+                                              file=file)])
 
     def _mirror_queue(self):
         name = config.mirror_queue.name
         return aws.sqs_queue(name)
 
-    def _queue_messages(self, messages: Iterable[MirrorAction]) -> int:
+    def _queue_actions(self, actions: Iterable[MirrorAction]) -> int:
         rate_limit = float(aws.sqs_fifo_rate_limit)
         if config.is_in_lambda:
             rate_limit /= config.mirroring_concurrency
         return self._queues.send_messages(self._mirror_queue(),
-                                          map(MirrorAction.to_sqs, messages),
+                                          map(MirrorAction.to_sqs, actions),
                                           rate_limit=rate_limit)
 
     #: Since we track the ETags of all parts of a multipart file in SQS
@@ -467,7 +467,7 @@ class MirrorService(BaseMirrorService, HasCachedHttpClient):
     _file_object_content_type = 'application/octet-stream'
 
     def mirror(self, action: MirrorAction):
-        self._queue_messages(self._mirror(action))
+        self._queue_actions(self._mirror(action))
 
     @singledispatchmethod
     def _mirror(self, a: MirrorAction):

--- a/src/azul/indexer/mirror_service.py
+++ b/src/azul/indexer/mirror_service.py
@@ -498,7 +498,6 @@ class MirrorService(BaseMirrorService, HasCachedHttpClient):
                  prefix.num_partitions, str(partitioned_source.spec), self.catalog)
 
         for partition in prefix.partition_prefixes():
-            log.debug('Queueing partition %r', partition)
             yield MirrorPartitionAction(catalog=self.catalog,
                                         source=partitioned_source,
                                         prefix=partition)
@@ -515,7 +514,6 @@ class MirrorService(BaseMirrorService, HasCachedHttpClient):
             assert file.size <= self.max_file_size, R(
                 'File too big', file, self.max_file_size)
             if self.may_mirror(file.size):
-                log.debug('Queueing file %r', file)
                 yield MirrorFileAction(catalog=self.catalog,
                                        source=a.source,
                                        prefix=a.prefix,
@@ -542,7 +540,6 @@ class MirrorService(BaseMirrorService, HasCachedHttpClient):
                 log.info('Mirroring file via multi-part upload: %r', a.file)
                 upload = self._create_upload(a.file)
                 next_part = self._mirror_first_part(a.file, upload)
-                log.info('Queueing part #%d of file %r', next_part.index, a.file)
                 yield MirrorPartAction(catalog=self.catalog,
                                        source=a.source,
                                        prefix=a.prefix,
@@ -610,7 +607,6 @@ class MirrorService(BaseMirrorService, HasCachedHttpClient):
                                      file=a.file,
                                      upload=upload)
         else:
-            log.info('Queueing part #%d of file %r', next_part.index, a.file)
             yield MirrorPartAction(catalog=self.catalog,
                                    source=a.source,
                                    prefix=a.prefix,

--- a/src/azul/queues.py
+++ b/src/azul/queues.py
@@ -48,6 +48,7 @@ from azul import (
 )
 from azul.attrs import (
     DiscriminatingPolymorphicSerializableAttrs,
+    SerializableAttrs,
 )
 from azul.deployment import (
     aws,
@@ -84,8 +85,8 @@ if TYPE_CHECKING:
     )
 
 
-@attrs.frozen(kw_only=True)
-class SQSMessage:
+@attrs.frozen(kw_only=True, repr=False)
+class SQSMessage(SerializableAttrs):
     body: JSON
 
     #: Approximate number of times this message has been received, or None if
@@ -110,6 +111,11 @@ class SQSMessage:
         return cls(id=json_str(record.to_dict()['messageId']),
                    body=json.loads(record.body),
                    attempts=int(json_str(attributes['ApproximateReceiveCount'])))
+
+    def __repr__(self) -> str:
+        # We profusely log instances of this class, so we would like their
+        # logged form to benefit from CloudWatch's automatic parsing of JSON
+        return json.dumps(self.to_json())
 
 
 @attrs.frozen(kw_only=True)
@@ -195,13 +201,16 @@ class Queues:
     def send_messages(self,
                       queue: 'Queue',
                       messages: Iterable[SQSMessage],
-                      rate_limit: float | None = None
+                      rate_limit: float | None = None,
+                      log_level: int = logging.DEBUG
                       ) -> int:
         num_messages = 0
         for batch in chunked(messages, self.batch_size):
             entries = [message.to_batch_entry(i) for i, message in enumerate(batch)]
             start = time.time()
             queue.send_messages(Entries=entries)
+            for message in batch:
+                self._log_queued_message(message, log_level)
             if rate_limit is not None:
                 period = 1 / rate_limit
                 time_spent = time.time() - start
@@ -212,8 +221,15 @@ class Queues:
             num_messages += len(batch)
         return num_messages
 
-    def send_message(self, queue: 'Queue', message: SQSMessage):
+    def send_message(self,
+                     queue: 'Queue',
+                     message: SQSMessage,
+                     log_level: int = logging.DEBUG):
         queue.send_message(**message.to_entry())
+        self._log_queued_message(message, log_level)
+
+    def _log_queued_message(self, message: SQSMessage, log_level: int):
+        log.log(log_level, 'Queued message %r', message)
 
     def _cleanup_messages(self, queue: 'Queue', messages: Iterable['Message']):
         message_batches = list(more_itertools.chunked(messages, self.batch_size))

